SEARCH-799 filtered results db

eliaporciani opened the pull request.

derekhulley commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -529,38 +529,21 @@ public void setPostProcessDenies(boolean postProcessDenies)
 
-    {
+
+    private ResultSet decidePermissions(ResultSet returnedObject, List<ConfigAttributeDefintion> supportedDefinitions){
```
formatting

derekhulley replied:
Braces new line

tunaaksoy commented on src/test/java/org/alfresco/repo/search/impl/querymodel/impl/db/DBQueryTest.java:
```diff
@@ -268,8 +264,9 @@ protected void createTestData() throws Exception
-        this.authenticationComponent.setSystemUserAsCurrentUser();
-        
+//        this.authenticationComponent.setSystemUserAsCurrentUser();
+        this.authenticationComponent.setCurrentUser("admin");
```
Why are we setting the current user as admin? What was the issue?

eliaporciani replied:
My mistake. I have put it back as before.

tunaaksoy commented on src/test/java/org/alfresco/repo/search/impl/querymodel/impl/db/DBQueryTest.java:
```diff
@@ -66,11 +66,7 @@
-import org.alfresco.service.cmr.search.ResultSetRow;
-import org.alfresco.service.cmr.search.SearchParameters;
-import org.alfresco.service.cmr.search.SearchService;
+import org.alfresco.service.cmr.search.*;
```
We don't do wildcard import. Can we please fix this?

eliaporciani replied:
Removed wildcard.

tunaaksoy commented:
I get the idea about the change but I was not able to review ACLEntryAfterInvocationProvider.java as the formatting makes really hard to see the changes. @eliaporciani is it possible to fix the formatting please?

tunaaksoy commented:
If I can read this correctly the reason for ACLEntryAfterInvocationProvider being shown changed so much is that some of the existing code has been extracted to helper methods for code reuse. I think the fix is OK but I'd like to get @michaelsuzukisagi's opinion on this as well.

tpage-alfresco commented:
Please could you update to add a brief overview of the approach taken in the description of this pull request?  Currently there is no description here, and the ticket seems to mention more than one possible approach.

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/WeakFilteringResultSet.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.BitSet;
+
+public class WeakFilteringResultSet extends FilteringResultSet {
```
Javadoc

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -529,38 +529,19 @@ public void setPostProcessDenies(boolean postProcessDenies)
-    private ResultSet decideNew(Authentication authentication, Object object, ConfigAttributeDefinition config, ResultSet returnedObject) throws AccessDeniedException
 
+
+    private ResultSet decidePermissions(ResultSet returnedObject, List<ConfigAttributeDefintion> supportedDefinitions)
```
Javadoc would be very helpful here to give some indication of what will happen in this large method.

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
         }
+        else
+        {
+            for (int i = 0; i < maxSize.intValue() && i < returnedObject.length(); i++)
```
Why are we calling `intValue` here?  Can't we just autobox it?

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
+            }
+            else
+            {
+                filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.UNLIMITED, PermissionEvaluationMode.EAGER, returnedObject.getResultSetMetaData()
```
This would be clearer if we refactor to something like:
```
LimitBy limitBy = (returnedObject.length > maxSize ? FINAL_SIZE : UNLIMITED);
filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(limitBy, ...));
```

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
-            maxSize = new Integer(returnedObject.getResultSetMetaData().getSearchParameters().getMaxItems());
+            maxSize = returnedObject.getResultSetMetaData().getSearchParameters().getMaxItems();
         }
         if ((maxSize == null) && (returnedObject.getResultSetMetaData().getSearchParameters().getLimitBy() == LimitBy.FINAL_SIZE))
```
Actually the whole of this method could be rewritten as something like:
```
if (params.getMaxItems() >= 0)
{
    return params.getMaxItems() + params.getSkipCount();
}
else if (params.getLimitBy() == FINAL_SIZE)
{
    return params.getLimit() + params.getSkipCount();
}
return null;
```

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
+    }
 
+    /**
+     * Get the mac size from a resultSet.
```
Also - it would be nice to expand on what "max size" really means.  I think this is an upper bound on the number of results that could be included in the result set?  What would cause the number of results to be less than the output from this method?

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
-        }
+        FilteringResultSet filteringResultSet = new FilteringResultSet(returnedObject);
 
         // record the start time
```
What happens if we exceed the maximum allowed execution time?  Does the user get nothing?

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
+                // If supportedDefinitions is not passed as parameter, it uses new method to permission filtering.
+                if (supportedDefinitions == null)
+                {
+                    if (permissionService.hasReadPermission(nodeRef) == AccessStatus.DENIED)
```
These two statements could be AND-ed together.

tpage-alfresco replied:
Oh - actually I didn't spot the `else` statement.  Instead we could invert the outside if statement to get:
```
if (supportedDefinitions != null)
{
    ...
}
else if (permissionService.hasReadPermission(nodeRef) == DENIED)
{
    ...
}
```

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
-            return null;
+            return returnedObject;
         }
+        else
```
No need for an else statement here.

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
 
-        long maxCheckTime = maxPermissionCheckTimeMillis;
-        if (returnedObject.getResultSetMetaData().getSearchParameters().getMaxPermissionCheckTimeMillis() >= 0)
+    private ResultSet decideNew(Authentication authentication, Object object, ConfigAttributeDefinition config, ResultSet returnedObject) throws AccessDeniedException
```
Javadoc

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
+    }
 
-        for (int i = 0; i < returnedObject.length(); i++)
+    private ResultSet decideOld(Authentication authentication, Object object, ConfigAttributeDefinition config, ResultSet returnedObject) throws AccessDeniedException
```
Javadoc.  What does "decideOld" mean compared with "decideNew"?  I can't think of anything that should be old and new here.

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
-                    testNodeRef = returnedObject.getChildAssocRef(i).getParentRef();
-                }
+        List<ConfigAttributeDefintion> supportedDefinitions = extractSupportedDefinitions(config);
+        ResultSet resultSet = supportedDefinitions.size() == 0? returnedObject: decidePermissions(returnedObject, supportedDefinitions);
```
Rather than `.size() == 0` we can use `.isEmpty()`, which looks a tiny bit neater.

tpage-alfresco commented on src/test/java/org/alfresco/repo/search/impl/querymodel/impl/db/DBQueryTest.java:
```diff
@@ -899,7 +900,36 @@ public void testAFTS()
+        int count = 4;
+        ResultSet results = query(SearchService.LANGUAGE_FTS_ALFRESCO, query, count);
+        assert(results.length() == count);
+        assert(results.getNumberFound() == numFolders);
```
Maybe consider using `assertEquals("Useful message", results.getNumberFound(), numFolders)`

tpage-alfresco commented on src/test/java/org/alfresco/repo/search/impl/querymodel/impl/db/DBQueryTest.java:
```diff
@@ -899,7 +900,36 @@ public void testAFTS()
     }
+
+    @Test
+    public void testAftsPagination()
```
Please could you add some javadoc explaining roughly what happens in this test?  E.g. "Check that by limiting the maxCount we can get four results returned from a total of six."

tpage-alfresco commented on src/test/java/org/alfresco/repo/search/impl/querymodel/impl/db/DBQueryTest.java:
```diff
@@ -899,7 +900,36 @@ public void testAFTS()
+        assert(results.getNumberFound() == numFolders);
+    }
     
+    public ResultSet query(String ql, String query, Integer limit)
```
It would be worth adding some Javadoc here too, otherwise no one will be able to reuse this method.

eliaporciani commented:
I added Javadoc and useful descriptions to methods. I have also applied all the changes you proposed and I refactored decideOld and decideNew methods. 
Some of the things in the original code are not so clear, so I left them as they were before the code refactoring.

tpage-alfresco commented on src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java:
```diff
@@ -654,95 +691,56 @@ else if (cad.typeString.equals(AFTER_ACL_PARENT))
+        if ((maxSize != null) && (searchParameters.getSkipCount() >= 0))
         {
-            maxSize = maxSize + returnedObject.getResultSetMetaData().getSearchParameters().getSkipCount();
+            maxSize = maxSize + searchParameters.getSkipCount();
```
You can still refactor this if statement out:
```
if (maxItems >=0)
{
    maxSize = maxItems + skipCount;
}
else if (limitBy >= 0)
{
    maxSize = limit + skipCount;
}
```

tpage-alfresco commented:
Looks much better, but this is the only thing I noticed was still there: https://github.com/Alfresco/alfresco-repository/pull/290#discussion_r250900825

tpage-alfresco commented:
LGTM :+1:
